File: 백준/Gold/1043. 거짓말/거짓말.java
class Main {
	
	private static int[] parent;
	private static int N;
	
	public static void main(String[] args) throws Exception {
		N = read();
		int M = read();
		int n = read();
		
		makeSet();
		
		for (int i = 0; i < n; i++) {
			union(0, read());
		}
		
		int[][] party = new int[M][];
		
		for (int i = 0; i < M; i++) {
			n = read();
			party[i] = new int[n];
			
			for (int j = 0; j < n; j++) {
				party[i][j] = read();
				union(party[i][0], party[i][j]);
			}
		}
		
		for (int i = 0; i < M; i++) {
			for (int j = 0; j < party[i].length; j++) {
				union(party[i][0], party[i][j]);
			}
		}
		
		int answer = M;
		
		for (int i = 0; i < M; i++) {
			for (int j = 0; j < party[i].length; j++) {
				if (parent[party[i][j]] > 0) {
					continue;
				}
				
				answer--;
				break;
			}
		}
		
		System.out.print(answer);
	}
	
	private static int read() throws Exception {
		int c, n = System.in.read() & 15;
		
		while ((c = System.in.read()) > 32) {
			n = (n << 3) + (n << 1) + (c & 15);
		}
		
		return n;
	}
	
	private static void makeSet() {
		parent = new int[N + 1];
		
		for (int i = 1; i <= N; i++) {
			parent[i] = i;
		}
	}
	
	private static void union(int x, int y) {
		int px = findSet(x);
		int py = findSet(y);
		
		if (px == py) {
			return;
		}
		
		if (px < py) {
			parent[py] = px;
			
		} else {
			parent[px] = py;
		}
	}
	
	private static int findSet(int x) {
		if (x == parent[x]) {
			return x;
		}
		
		return parent[x] = findSet(parent[x]);
	}
}

File: 백준/Gold/1043. 거짓말/거짓말.py
import sys
input = sys.stdin.readline

def find(idx):
    if parent[idx] != idx:
        parent[idx] = find(parent[idx])
    return parent[idx]

def union(i,j):
    a = find(i)
    b = find(j)
    parent[max(a,b)] = min(a,b)

def check(i,j):
    return find(i) == find(j)


n,m = map(int, input().split())
knows = [False]*(n+1)
parent = [i for i in range(n+1)]
t, *tp = map(int, input().split())
for i in range(t):
    knows[tp[i]] = True

parties = []
for _ in range(m):
    p,*pp = map(int, input().split())
    parties.append(pp)

    if p>1:
        for i in range(p-1):
            union(pp[i],pp[i+1])

for i in range(1,n+1):
    if knows[i]:
        knows[find(i)] = True

if t == 0:
    print(m)
else:
    res = 0
    for party in parties:
        flag = True
        for p in party:
            if knows[find(p)]:
                flag = False
        if flag:
            res +=1
    print(res)